### ISSUE TITLE
Characters disappear from the forms as typing

### ISSUE THREAD

**Issue by adalbus** [rajeshpillai/react-dynamic-form#1]
I am having an issue when typing on the input textboxes. Whatever I type, it resets after every keystroke. 

I attached a debugger to your project. I can see that onChange is fired and it keeps the state for a while but later on the event pipeline, the state is reset (erasing the inputs). 

I found reset happens inside one of the react libraries.

....\node_modules\react-dom\cjs\react-dom.development.js
      // Synchronously dispatch our fake event. If the user-provided function
      // errors, it will trigger our global error handler.
      evt.initEvent(evtType, false, false);
      fakeNode.dispatchEvent(evt);

I did git clone on your repo, and I tested on Chrome and Edge.

I will appreciate any pointers.

**Comment by rcmgleite** [comment]
👍

**Comment by bsvveen** [comment]
Following the recommendations in: https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html i removed _getDerivedStateFromProps_ entirely and in renderForm: 

replaced: `let target = key;  
            value = this.state[target];`

with

`
let defaultValue = this.props.defaultValues[m.key] || "";
let value = this.state[key] || defaultValue; 
`

**Comment by mvd5** [comment]
I'm a web-programming beginner having the same issue (key stokes do not update the form); bsevven's recommendation is a great place to start but I couldn't get it to resolve the issue.  I'll post back if I figure out a solution.  Note that the "edit" button on the table DOES successfully load data into the form, you just cannot edit it yourself ... [React 16.6.3 and Node 8.10.0]

**Comment by mvd5** [comment]
Ok, to fix this issue follow bsvveen's post and change the following:
**original: onChange={(e)=>{this.onChange(e, defaultValues)}}
new: onChange={(e)=>{this.onChange(e, key)}}**
This passes in the actual key, not an object, so that setState changes the right entry.
Now, there are still errors in selecting the skill and hitting the submit button.  I'll look into those next.

**Comment by mvd5** [comment]
Ok, the solutions to the two subsequent errors above are below.  they are by no means elegant, just checks for NULL basically to satisfy the base-case:
In App.js, check for null and just return a blank string if not defined:
**<td>{d.skills ? d.skills.join(",") : ""}</td>**

In the component: index.js; this is from onChange() : check for NULL before distributing the other checkbox selections:
                **if (this.state[key]) {
                    this.setState({
                        [key]: [e.target.value, ...this.state[key]]
                    });     
                } else {
                    this.setState({
                        [key]: [e.target.value]
                    });**
                }

**Comment by Rafahss** [comment]
@bsvveen 
I did what you said, but I still have the same problem. Can you help me?

**Comment by mvd5** [comment]
The above is hard to follow, certainly.  I'll try to help - can you post your code or more detail?  Or should this go offline..

**Comment by Rafahss** [comment]
@mvd5 
Thanks for the answer.
I am using the code that is in the same github, still had not changed anything, however the behavior is totally strange.

**Comment by mvd5** [comment]
Ok trying to remember myself here.  We are in DynamicForm.js.  We've commented out the "getDerivedStateFromProps" function as was recommended on another post.  We are editing "renderForm":
Gee .. I needed some edits to make the form do what I wanted and I cannot post it all here unfortunately.  I remember I was solving a couple of irritations: text input which wouldn't appear and when I deleted the last numeric digit (like deleting a zero) it was re-appear in the "number" type; which is technically correct but irritating to the user.
I solved this broadly by creating a shadow copy of the JSON input (calling it "defaults") and compared what the user provided initially to what the form was updating at the moment.  I had to make a distinction in render() between a "first time" render and a re-render because the user just entered a single character into a text box.
This is probably not very helpful, but maybe provides you an idea.  So, put a tracer console.log in render() to see when it updates and what the value is when you render first time vs an update.  You maybe surprised to see what it is doing.

**Comment by mvd5** [comment]
Oh, then, in the "renderForm()" function - check if this is a "first time" render or a re-render because of a single user key press ...
